### ISSUE TITLE
Introduce support for maven wrapper

### ISSUE THREAD

**Issue by fy-gualtiero-testa** [eclipse-m2e/m2e-core#617]
It would be useful to have m2e detect the presence of the maven wrapper installation in the project and use that version to build the project.
It could be an option in the lanch configuration editor like "Use maven wrapper if available"

The main benefit is to have Eclipse builds consistent with the builds launched from the command line.

**Comment by laeubi** [comment]
> The main benefit is to have Eclipse builds consistent with the builds launched from the command line.

You mean `Run As > Maven`? The Eclipse build what is running *inside* Eclipse as you edit your source files.

**Comment by fy-gualtiero-testa** [comment]
> You mean `Run As > Maven`? The Eclipse build what is running _inside_ Eclipse as you edit your source files.
No, Maven Wrapper (https://maven.apache.org/wrapper/) is a tool which creates a project specific Maven installation. Having Maven version defined at project level helps to be consistent among team members, CI tools and IDEs because they all use the same Maven version.
Maven wrapper creates two script files (mvnw and mvnw.bat) and a .mvn dir in the project root dir; the scripts can be used as a replacement for the Maven mvn command.

I mean
`   ./mvnw clean verify`
instead of
`  mvn clean verify`
  
This issue is to have Eclipse check the presence of the Maven wrapper files in the project root dir and, if they are present, use the script files instead of the configured Maven installation.

Benefits:
1. No need to configure specific Maven versions in Eclipse
2. Every project can use its own Maven version
3. Maven runs from Eclipse (Run as --> maven) are consistent with command line runs.

**Comment by HannesWell** [comment]
Considering the Maven-Wrapper when launching a Maven-build should be straight forward.
In the end it is a external Maven-installation, which is already supported. M2E would only have to be taught to use that automatically. Probably somewhere in the vicinity of `org.eclipse.m2e.core.internal.launch.MavenExternalRuntime` and in `org.eclipse.m2e.launching`.

Using the Maven-Wrapper installation when building a project in the IDE will likely be more complicated. Theoretically we could generate a suitable Manifest.MF and install such wrapper-maven-runtime bundle into the running Eclipse at runtime. It then has to be ensured that the M2E bundles are wired to that bundle or the build would have to manually search for a fitting maven-runtime-bundle and load the classes from there 'manually'. Which could theoretically be done per project (at with the latest work of @laeubi and @mickaelistria which aims to allow 'individual' mavens per project, but at the moment this is intended to be always the same maven version and bundle).

But M2E plug-ins use the Maven-classes too so we would have to ensure that M2E works with a wide variety of Maven versions (supporting always the latest version does not necessarily mean that all previous versions work automatically).

**Comment by laeubi** [comment]
m2e always uses the "embedded" runtime in eclipse and I don't see a problem here, the purpose of the wrapper is that one don't need (that particular version) installed on your machine and it is downloaded automatically, so m2e is already a "mvn wrapper" like :-)

So I think this should really be limited to running the build, and actually we probably simply adjust the runner to run the commandline `/mvnw` instead of the `java -jar ....`

**Comment by mickaelistria** [comment]
Usage of embedded Maven vs Maven wrapper is far from being the main reason of inconsistencies of CLI vs m2e builds. The main reasons are support of non-m2e-stamped mojos, support of extensions, support of maven.conf and so on; the wrapper has a neglictable effect on outcome compared to those.
So I wouldn't advise anyone spending time on this particular issue, until other ones listed above have much improved.
